File: python/lan_device_manager.py
import csv
import os
import re
import sys
from typing import Dict, List, Optional


class DeviceManager:
    """Device manager using CSV as the persisted database."""

    def __init__(self, csv_path: str = "data/lan_devices.csv"):
        self.csv_path = csv_path
        self.ensure_file_exists()

    def ensure_file_exists(self):
        directory = os.path.dirname(self.csv_path)
        if directory and not os.path.exists(directory):
            os.makedirs(directory, exist_ok=True)
        if not os.path.exists(self.csv_path):
            with open(self.csv_path, "w", encoding="utf-8", newline="") as file:
                writer = csv.writer(file)
                writer.writerow([
                    "device_id",
                    "name",
                    "device_type",
                    "ip",
                    "mac",
                    "status",
                    "location",
                ])

    def load_devices(self) -> List[Dict[str, str]]:
        devices: List[Dict[str, str]] = []
        if not os.path.exists(self.csv_path):
            self.ensure_file_exists()

        try:
            with open(self.csv_path, "r", encoding="utf-8", newline="") as file:
                reader = csv.DictReader(file)
                for row in reader:
                    if not row:
                        continue
                    devices.append({
                        "device_id": (row.get("device_id") or "").strip(),
                        "name": (row.get("name") or "").strip(),
                        "device_type": (row.get("device_type") or "").strip(),
                        "ip": (row.get("ip") or "").strip(),
                        "mac": (row.get("mac") or "").strip(),
                        "status": (row.get("status") or "").strip(),
                        "location": (row.get("location") or "").strip(),
                    })
        except FileNotFoundError:
            self.ensure_file_exists()
        return devices

    def save_devices(self, devices: List[Dict[str, str]]):
        with open(self.csv_path, "w", encoding="utf-8", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["device_id", "name", "device_type", "ip", "mac", "status", "location"])
            for device in devices:
                writer.writerow([
                    device.get("device_id", ""),
                    device.get("name", ""),
                    device.get("device_type", ""),
                    device.get("ip", ""),
                    device.get("mac", ""),
                    device.get("status", ""),
                    device.get("location", ""),
                ])

    def validate_ipv4(self, ip: str) -> bool:
        pattern = r"^(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}$"
        return bool(re.fullmatch(pattern, ip))

    def normalize_mac(self, mac: str) -> str:
        sanitized = re.sub(r"[^0-9A-Fa-f]", "", mac)
        if len(sanitized) != 12:
            return ""
        chunks = [sanitized[i:i + 2] for i in range(0, 12, 2)]
        return "-".join(chunks).upper()

    def validate_mac(self, mac: str) -> bool:
        if not mac:
            return False
        normalized = self.normalize_mac(mac)
        return bool(normalized) and bool(re.fullmatch(r"(?:[0-9A-Fa-f]{2}-){5}[0-9A-Fa-f]{2}", normalized))

    def find_device(self, device_id: str) -> Optional[Dict[str, str]]:
        for device in self.load_devices():
            if device["device_id"].lower() == device_id.lower():
                return device
        return None

    def add_device(self, device: Dict[str, str]) -> str:
        devices = self.load_devices()
        if any(d["device_id"].lower() == device["device_id"].lower() for d in devices):
            return "设备ID已存在，不能重复添加。"
        if any(d["ip"] == device["ip"] for d in devices):
            return "IP地址冲突：该 IP 已被其他设备使用。"
        if any(d["mac"].upper() == device["mac"].upper() for d in devices):
            return "MAC地址冲突：该 MAC 已被其他设备使用。"
        if not self.validate_ipv4(device["ip"]):
            return "IP地址格式不正确，请输入合法 IPv4 地址。"
        if not self.validate_mac(device["mac"]):
            return "MAC地址格式不正确，请使用 XX-XX-XX-XX-XX-XX 格式。"
        devices.append(device)
        self.save_devices(devices)
        return "添加成功。"

    def delete_device(self, device_id: str) -> str:
        devices = self.load_devices()
        for index, device in enumerate(devices):
            if device["device_id"].lower() == device_id.lower():
                del devices[index]
                self.save_devices(devices)
                return f"设备 {device_id} 已删除。"
        return "未找到该设备。"

    def update_device(self, device_id: str, new_device: Dict[str, str]) -> str:
        devices = self.load_devices()
        for index, device in enumerate(devices):
            if device["device_id"].lower() == device_id.lower():
                if not self.validate_ipv4(new_device["ip"]):
                    return "IP地址格式不正确，请输入合法 IPv4 地址。"
                if not self.validate_mac(new_device["mac"]):
                    return "MAC地址格式不正确，请使用 XX-XX-XX-XX-XX-XX 格式。"
                for other in devices:
                    if other is device:
                        continue
                    if other["ip"] == new_device["ip"]:
                        return "IP地址冲突：该 IP 已被其他设备使用。"
                    if other["mac"].upper() == new_device["mac"].upper():
                        return "MAC地址冲突：该 MAC 已被其他设备使用。"
                devices[index] = new_device
                self.save_devices(devices)
                return "更新成功。"
        return "未找到该设备。"

    def query_device(self, keyword: str) -> List[Dict[str, str]]:
        devices = self.load_devices()
        result: List[Dict[str, str]] = []
        keyword_lower = keyword.lower()
        for device in devices:
            if (
                keyword_lower in device["device_id"].lower()
                or keyword_lower in device["name"].lower()
                or keyword_lower in device["device_type"].lower()
                or keyword_lower in device["ip"].lower()
                or keyword_lower in device["mac"].lower()
                or keyword_lower in device["status"].lower()
                or keyword_lower in device["location"].lower()
            ):
                result.append(device)
        return result

    def print_devices(self, devices: List[Dict[str, str]]):
        if not devices:
            print("没有找到设备记录。")
            return

        print("-" * 132)
        print(f"{'设备ID':<10} | {'名称':<12} | {'类型':<12} | {'IP':<16} | {'MAC':<20} | {'状态':<8} | {'位置':<12}")
        print("-" * 132)
        for d in devices:
            print(f"{d['device_id']:<10} | {d['name']:<12} | {d['device_type']:<12} | {d['ip']:<16} | {d['mac']:<20} | {d['status']:<8} | {d['location']:<12}")
        print("-" * 132)


def print_menu():
    print("\n===== 局域网设备管理系统 =====")
    print("1. 添加设备")
    print("2. 删除设备")
    print("3. 修改设备")
    print("4. 查询设备")
    print("5. 显示全部设备")
    print("0. 退出")


def input_device_data() -> Dict[str, str]:
    device = {
        "device_id": input("请输入设备ID: ").strip(),
        "name": input("请输入设备名称: ").strip(),
        "device_type": input("请输入设备类型: ").strip(),
        "ip": input("请输入IP地址: ").strip(),
        "mac": input("请输入MAC地址(XX-XX-XX-XX-XX-XX): ").strip(),
        "status": input("请输入状态(在线/离线/维护中): ").strip(),
        "location": input("请输入位置: ").strip(),
    }
    return device


def main_menu():
    while True:
        print_menu()
        choice = input("请选择功能：").strip()
        manager = DeviceManager()

        if choice == "1":
            device = input_device_data()
            print(manager.add_device(device))
        elif choice == "2":
            device_id = input("请输入要删除的设备ID: ").strip()
            print(manager.delete_device(device_id))
        elif choice == "3":
            device_id = input("请输入要修改的设备ID: ").strip()
            if manager.find_device(device_id) is None:
                print("未找到该设备。")
                continue
            new_device = input_device_data()
            print(manager.update_device(device_id, new_device))
        elif choice == "4":
            keyword = input("请输入查询关键字: ").strip()
            result = manager.query_device(keyword)
            manager.print_devices(result)
        elif choice == "5":
            manager.print_devices(manager.load_devices())
        elif choice == "0":
            print("程序已退出。")
            exit(0)
        else:
            print("无效选项，请重试。")


if __name__ == "__main__":
    try:
        main_menu()
    except KeyboardInterrupt:
        print("\n程序已中断。")
        sys.exit(0)

File: c/lan_device_manager.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <windows.h>

#define MAX_DEVICES 200
#define MAX_LINE 256

struct Device {
    char device_id[32];
    char name[64];
    char device_type[32];
    char ip[32];
    char mac[32];
    char status[32];
    char location[64];
};

static void set_utf8_console(void) {
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);
    setlocale(LC_ALL, ".65001");
}

static int load_devices(struct Device devices[], const char *path, int *count) {
    FILE *fp = fopen(path, "r");
    char line[MAX_LINE];
    char *token;
    int index = 0;

    if (fp == NULL) {
        fp = fopen(path, "w+");
        if (fp == NULL) {
            printf("无法创建数据文件。\n");
            return 0;
        }
        fprintf(fp, "device_id,name,device_type,ip,mac,status,location\n");
        fclose(fp);
        fp = fopen(path, "r");
    }

    fgets(line, sizeof(line), fp);
    while (fgets(line, sizeof(line), fp) != NULL && index < MAX_DEVICES) {
        char buffer[MAX_LINE];
        strncpy(buffer, line, sizeof(buffer) - 1);
        buffer[sizeof(buffer) - 1] = '\0';

        token = strtok(buffer, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].device_id, sizeof(devices[index].device_id), "%s", token);

        token = strtok(NULL, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].name, sizeof(devices[index].name), "%s", token);

        token = strtok(NULL, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].device_type, sizeof(devices[index].device_type), "%s", token);

        token = strtok(NULL, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].ip, sizeof(devices[index].ip), "%s", token);

        token = strtok(NULL, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].mac, sizeof(devices[index].mac), "%s", token);

        token = strtok(NULL, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].status, sizeof(devices[index].status), "%s", token);

        token = strtok(NULL, ",\r\n");
        if (token == NULL) continue;
        snprintf(devices[index].location, sizeof(devices[index].location), "%s", token);

        index++;
    }

    fclose(fp);
    *count = index;
    return 1;
}

static int save_devices(struct Device devices[], int count, const char *path) {
    FILE *fp = fopen(path, "w");
    if (fp == NULL) {
        printf("无法写入数据文件。\n");
        return 0;
    }

    fprintf(fp, "device_id,name,device_type,ip,mac,status,location\n");
    for (int i = 0; i < count; i++) {
        fprintf(fp, "%s,%s,%s,%s,%s,%s,%s\n",
                devices[i].device_id,
                devices[i].name,
                devices[i].device_type,
                devices[i].ip,
                devices[i].mac,
                devices[i].status,
                devices[i].location);
    }

    fclose(fp);
    return 1;
}

static int validate_ipv4(const char *ip) {
    int a, b, c, d;
    if (sscanf(ip, "%d.%d.%d.%d", &a, &b, &c, &d) != 4) return 0;
    if (a < 0 || a > 255 || b < 0 || b > 255 || c < 0 || c > 255 || d < 0 || d > 255) return 0;
    return 1;
}

static int validate_mac(const char *mac) {
    int len = (int)strlen(mac);
    if (len != 17) return 0;
    for (int i = 0; i < len; i++) {
        if (i % 3 == 2) {
            if (mac[i] != '-') return 0;
        } else if (!((mac[i] >= '0' && mac[i] <= '9') || (mac[i] >= 'A' && mac[i] <= 'F') || (mac[i] >= 'a' && mac[i] <= 'f'))) {
            return 0;
        }
    }
    return 1;
}

static int has_ip_conflict(struct Device devices[], int count, const char *ip, int skip_index) {
    for (int i = 0; i < count; i++) {
        if (i == skip_index) continue;
        if (strcmp(devices[i].ip, ip) == 0) return 1;
    }
    return 0;
}

static int has_mac_conflict(struct Device devices[], int count, const char *mac, int skip_index) {
    for (int i = 0; i < count; i++) {
        if (i == skip_index) continue;
        if (strcmp(devices[i].mac, mac) == 0) return 1;
    }
    return 0;
}

static void add_device(struct Device devices[], int *count) {
    struct Device new_device;
    printf("请输入设备ID: ");
    scanf(" %31s", new_device.device_id);
    printf("请输入名称: ");
    scanf(" %63s", new_device.name);
    printf("请输入类型: ");
    scanf(" %31s", new_device.device_type);
    printf("请输入IP地址: ");
    scanf(" %31s", new_device.ip);
    printf("请输入MAC地址(XX-XX-XX-XX-XX-XX): ");
    scanf(" %31s", new_device.mac);
    printf("请输入状态: ");
    scanf(" %31s", new_device.status);
    printf("请输入位置: ");
    scanf(" %63s", new_device.location);

    if (!validate_ipv4(new_device.ip)) {
        printf("IP地址格式不正确。\n");
        return;
    }
    if (!validate_mac(new_device.mac)) {
        printf("MAC地址格式不正确。\n");
        return;
    }
    if (has_ip_conflict(devices, *count, new_device.ip, -1)) {
        printf("IP地址冲突：该 IP 已被其他设备使用。\n");
        return;
    }
    if (has_mac_conflict(devices, *count, new_device.mac, -1)) {
        printf("MAC地址冲突：该 MAC 已被其他设备使用。\n");
        return;
    }

    devices[*count] = new_device;
    (*count)++;
    printf("添加成功。\n");
}

static void delete_device(struct Device devices[], int *count) {
    char device_id[32];
    int index = -1;
    printf("请输入要删除的设备ID: ");
    scanf(" %31s", device_id);

    for (int i = 0; i < *count; i++) {
        if (strcmp(devices[i].device_id, device_id) == 0) {
            index = i;
            break;
        }
    }

    if (index == -1) {
        printf("未找到该设备。\n");
        return;
    }

    for (int i = index; i < *count - 1; i++) {
        devices[i] = devices[i + 1];
    }
    (*count)--;
    printf("删除成功。\n");
}

static void update_device(struct Device devices[], int count) {
    char device_id[32];
    int index = -1;
    printf("请输入要修改的设备ID: ");
    scanf(" %31s", device_id);

    for (int i = 0; i < count; i++) {
        if (strcmp(devices[i].device_id, device_id) == 0) {
            index = i;
            break;
        }
    }

    if (index == -1) {
        printf("未找到该设备。\n");
        return;
    }

    printf("请输入新的名称: ");
    scanf(" %63s", devices[index].name);
    printf("请输入新的类型: ");
    scanf(" %31s", devices[index].device_type);
    printf("请输入新的IP地址: ");
    scanf(" %31s", devices[index].ip);
    printf("请输入新的MAC地址(XX-XX-XX-XX-XX-XX): ");
    scanf(" %31s", devices[index].mac);
    printf("请输入新的状态: ");
    scanf(" %31s", devices[index].status);
    printf("请输入新的位置: ");
    scanf(" %63s", devices[index].location);

    if (!validate_ipv4(devices[index].ip)) {
        printf("IP地址格式不正确。\n");
        return;
    }
    if (!validate_mac(devices[index].mac)) {
        printf("MAC地址格式不正确。\n");
        return;
    }
    if (has_ip_conflict(devices, count, devices[index].ip, index)) {
        printf("IP地址冲突：该 IP 已被其他设备使用。\n");
        return;
    }
    if (has_mac_conflict(devices, count, devices[index].mac, index)) {
        printf("MAC地址冲突：该 MAC 已被其他设备使用。\n");
        return;
    }

    printf("修改成功。\n");
}

static void query_device(struct Device devices[], int count) {
    char keyword[64];
    printf("请输入查询关键字: ");
    scanf(" %63s", keyword);

    int found = 0;
    for (int i = 0; i < count; i++) {
        if (strstr(devices[i].device_id, keyword) != NULL ||
            strstr(devices[i].name, keyword) != NULL ||
            strstr(devices[i].device_type, keyword) != NULL ||
            strstr(devices[i].ip, keyword) != NULL ||
            strstr(devices[i].mac, keyword) != NULL ||
            strstr(devices[i].status, keyword) != NULL ||
            strstr(devices[i].location, keyword) != NULL) {
            printf("%s | %s | %s | %s | %s | %s | %s\n",
                   devices[i].device_id,
                   devices[i].name,
                   devices[i].device_type,
                   devices[i].ip,
                   devices[i].mac,
                   devices[i].status,
                   devices[i].location);
            found = 1;
        }
    }

    if (!found) {
        printf("没有找到匹配设备。\n");
    }
}

static void display_all_devices(struct Device devices[], int count) {
    printf("----------------------------------------------------------------------------------------------------\n");
    printf("| %-10s | %-12s | %-12s | %-16s | %-20s | %-10s | %-12s |\n",
           "device_id", "name", "device_type", "ip", "mac", "status", "location");
    printf("----------------------------------------------------------------------------------------------------\n");
    for (int i = 0; i < count; i++) {
        printf("| %-10s | %-12s | %-12s | %-16s | %-20s | %-10s | %-12s |\n",
               devices[i].device_id,
               devices[i].name,
               devices[i].device_type,
               devices[i].ip,
               devices[i].mac,
               devices[i].status,
               devices[i].location);
    }
    printf("----------------------------------------------------------------------------------------------------\n");
}

static void show_menu(void) {
    printf("\n===== 局域网设备管理系统 =====\n");
    printf("1. 添加设备\n");
    printf("2. 删除设备\n");
    printf("3. 修改设备\n");
    printf("4. 查询设备\n");
    printf("5. 显示全部设备\n");
    printf("0. 退出\n");
}

int main(void) {
    struct Device devices[MAX_DEVICES];
    int count = 0;
    int choice;
    const char *path = "data/lan_devices.csv";

    set_utf8_console();
    load_devices(devices, path, &count);

    while (1) {
        show_menu();
        printf("请选择功能：");
        if (scanf("%d", &choice) != 1) {
            printf("输入错误，请重新输入。\n");
            while (getchar() != '\n');
            continue;
        }

        switch (choice) {
            case 1:
                add_device(devices, &count);
                save_devices(devices, count, path);
                break;
            case 2:
                delete_device(devices, &count);
                save_devices(devices, count, path);
                break;
            case 3:
                update_device(devices, count);
                save_devices(devices, count, path);
                break;
            case 4:
                query_device(devices, count);
                break;
            case 5:
                display_all_devices(devices, count);
                break;
            case 0:
                printf("程序已退出。\n");
                return 0;
            default:
                printf("无效选项。\n");
                break;
        }
    }

    return 0;
}
